presto-hive/src/main/java/com/facebook/presto/hive/StaticHiveCluster.java

```diff
@@ -14,35 +14,82 @@
 package com.facebook.presto.hive;
 
 import com.google.common.net.HostAndPort;
+import io.airlift.log.Logger;
 import org.apache.thrift.transport.TTransportException;
 
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.ThreadLocalRandom;
+
+import static com.facebook.presto.hive.RetryDriver.retry;
 import static com.google.common.base.Preconditions.checkArgument;
 import static com.google.common.base.Preconditions.checkNotNull;
 
 public class StaticHiveCluster
         implements HiveCluster
 {
-    private final HostAndPort address;
+    private static final Logger log = Logger.get(StaticHiveCluster.class);
+    private final List<HostAndPort> addresses;
     private final HiveMetastoreClientFactory clientFactory;
 
-    public StaticHiveCluster(HostAndPort address, HiveMetastoreClientFactory clientFactory)
+    public StaticHiveCluster(List<HostAndPort> addresses, HiveMetastoreClientFactory clientFactory)
     {
-        checkNotNull(address, "address is null");
-        checkArgument(address.hasPort(), "address does not have a port");
+        checkNotNull(addresses, "addresses is null");
+        checkArgument(addresses.size() > 0, "must have one or more addresses");
+        addresses.forEach(StaticHiveCluster::checkPort);
         checkNotNull(clientFactory, "clientFactory is null");
 
-        this.address = address;
+        this.addresses = addresses;
         this.clientFactory = clientFactory;
     }
 
+    private static void checkPort(HostAndPort address)
+    {
+        checkArgument(address.hasPort(), "address " + address + " does not have a port");
+    }
+
+    /**
+     * As per Hive HA metastore behavior, return the first metastore in the list of available
+     * metastores if a connection can be made, else randomly select one of the other metastores to try.
+     * Continue until a connection can be made, or until it times out.
+     */
     @Override
     public HiveMetastoreClient createMetastoreClient()
     {
         try {
-            return clientFactory.create(address.getHostText(), address.getPort());
+            return defaultMetastoreClient();
         }
         catch (TTransportException e) {
-            throw new RuntimeException("Failed connecting to Hive metastore: " + address, e);
+            if (addresses.size() > 1) {
```

Comment:
I don't think we need to special case here. The loop in `DiscoveryLocatedHiveCluster` should work if you order the list of addresses correctly.